### ISSUE TITLE
fish_vcs_prompt: improve format string idea and document it

### ISSUE THREAD

**Issue by ammgws** [fish-shell/fish-shell#7047]
As mentioned in #7033, `fish_git_prompt` can be supplied a string to format its output.

> This is little known, but fish_git_prompt (that's the name in fish 3.1, the old one with the dunders is kept for compatibility) takes a "format string" (this was at one point supposed to do more but didn't really pan out). It defaults to (%s), so if you want to leave off the parens just call fish_git_prompt %s.

This feature is undocumented, however it is quite rudimentary and IIRC not available in the svn or hg prompts, so before documenting it someone may want to work on those points.

**Comment by faho** [comment]
Yeah, I have a working prototype. My question is what this *should* include?

As I currently have it, you'd give a list of words for what you want to have where, like

```fish
fish_git_prompt branch dirty upstream
```

and that would then print the branch, the dirty-state and the upstream-state. Any unrecognized word is currently printed as-is, tho I think we want to ignore anything that looks like a word to make it more cross-compatible across versions.

The code for this is actually acceptable, it's just a little weird that it accepts the arguments as a list yet won't print a space between them.

Also I'm not sold on the need to reorder these things, and I have not found a nice way to specify e.g. "only print a separator here if there is something".

It looks like this:

![Screenshot_20200527_180133](https://user-images.githubusercontent.com/5185367/83044184-20660a80-a044-11ea-8268-98eedbe0a41d.png)

**Comment by shlemas** [comment]
Just to throw an idea out, what if fish provided something like:

```
function fish_git_prompt_custom -a custom_func
    # Gather all the git repo data, etc.
    ...
    eval "$custom_func $branch $upstream $dirty"
end
```

And then users could do something like:

```
function git_formatter -a branch upstream dirty
    # I can arbitrarily format my own string with the data passed in by fish_get_prompt_custom.
    echo "[$branch] ($upstream | $dirty)"
end

function fish_prompt
    printf "%s %s" (pwd) (fish_git_prompt_custom git_formatter)
end
```

**Comment by ariasuni** [comment]
I’m not sure if this is relevant but I find it annoying to not be able to reorder dirtystate/stagedstate/stashstate/untrackedfile/invalidstate. Maybe we should be able to use something like this?
https://github.com/fish-shell/fish-shell/blob/f75cf5c16bf8fa43ba99180a8e4ba008d5ffc685/share/functions/fish_git_prompt.fish#L513

I’m not sure where you want to draw the line between what should be available in fish out-of-the-box, and what should be be let to more advanced (sometimes cross-shell) prompts like [silver](https://github.com/reujab/silver) or [starship](https://starship.rs/).

**Comment by faho** [comment]
>Maybe we should be able to use something like this?

So presumably all that needs to be done is to only use that value as the default?

**Comment by ariasuni** [comment]
I’m not really sure. Now that the order has been changed (in https://github.com/fish-shell/fish-shell/commit/71166274a22f65be8f345bcf0f2d23f4136c36c2), I don’t need this feature.

I guess it would be useful to know who:
- Uses `fish_git_prompt`, and would like these configuration options
- Uses another solution, but would prefer using `fish_git_prompt` if it had these configuration options

Because advanced users may also want to reorder branch/statuses/special states (ongoing merge, rebase, etc.), but you if you offer to configure this, a format string makes it hard to handles spaces properly when one item is not displayed because irrelevant, etc.

**Comment by faho** [comment]
Oh, what I meant was to allow that variable to work we just need to do

```fish
set -q __fish_git_prompt_status_order
or set -g __fish_git_prompt_status_order ....
```

This used to set the variable unconditionally when `fish_git_prompt` was loaded, so if you set that in config.fish it would be overwritten.

But I've then noticed that we weren't actually using it in the non-informative mode.

**Comment by wesruv** [comment]
I'm not entirely sure I understand this ticket, but I think this might be helpful feedback?

I've been looking for how to remove the ` (` prefix and `)` suffix in `fish_git_prompt` for some time now, and I'm very thankful for the top comment in this issue. Using `%s` after `fish_git_prompt` does exactly what I wanted.

The why of `printf (fish_git_prompt %s)` or what `%s` means is completely opaque to me. Now that I have the answer, and I looked back at the docs, I'm not sure where I would have figured it out, or how to figure out what `%s` is about.

For context, I'm an:
* intermediate CLI user
* a copy paste monkey when it comes to fish config
* was able to figure out how to write some custom fish functions
* professional web developer by day (professional HTML/CSS/JS coder)
